### PR TITLE
Override `os._Environ.__ior__`

### LINKED ISSUE #6919: `os.environ |= dict[str, str]` results in `Incompatible types in assignment (expression has type "Dict[str, str]", variable has type "_Environ[str]")`
``` python
# test_case.py
from __future__ import annotations
import os
moar_env: dict[str, str] = {}
os.environ.update(moar_env)  # <-- copacetic
os.environ |= moar_env  # <-- doesn't pass validation
```

``` sh
% mypy --config=/dev/null test_case.py
/dev/null: No [mypy] section in config file
test_case.py:6: error: Incompatible types in assignment (expression has type "Dict[str, str]", variable has type "_Environ[str]")
Found 1 error in 1 file (checked 1 source file)
% mypy --version
mypy 0.931
% python --version
Python 3.9.9
```

### PR DESCRIPTION
**Pull request by posita**
Supporting the `|=` operator for `os.environ` was introduced in Python 3.9. However,
this led to false negatives in type checking statements like `os.environ |= {"k": "v"}`
and `os.environ |= [("k", "v")]` because the definition inherited from
`typing.MutableMapping` was insufficient. This addresses that deficiency.

Fixes #6919.

### CODE REVIEW

**Comment by posita** [comment]
I'm a little out of my depth here and not quite sure if/how to create a specific test case for this, but I did test the following by hand before and after this change:

``` python
from os import environ

# These now properly pass (but did not before)
environ |= {"a": "b"}
environ |= (("a", "c"),)
environ |= [("a", "d")]

# Properly caught (Argument 1 to "__ior__" of "_Environ" has incompatible type "Tuple[List[str]]"; expected "Iterable[Tuple[str, str]]" [arg-type])
environ |= (["a", "c"],)

# Properly caught (List item 0 has incompatible type "List[str]"; expected "Tuple[str, str]" [list-item])
environ |= [["a", "d"]]
```

File: stdlib/os/__init__.pyi [hunk 1]
```diff
@@ -245,6 +246,13 @@ class _Environ(MutableMapping[AnyStr, AnyStr], Generic[AnyStr]):
     def __setitem__(self, key: AnyStr, value: AnyStr) -> None: ...
     def __iter__(self) -> Iterator[AnyStr]: ...
     def __len__(self) -> int: ...
+    if sys.version_info >= (3, 9):
+        # We use @overload instead of a Union for reasons similar to those given for
+        # overloading MutableMapping.update in stdlib/typing.pyi
+        @overload
+        def __ior__(self, __value: Mapping[AnyStr, AnyStr]) -> Self: ...
+        @overload
+        def __ior__(self, __value: Iterable[Tuple[AnyStr, AnyStr]]) -> Self: ...
```
Comment (JelleZijlstra) [inline stdlib/os/__init__.pyi:255]:
```suggestion
        def __ior__(self: Self, __value: Mapping[AnyStr, AnyStr]) -> Self: ...
        @overload
        def __ior__(self: Self, __value: Iterable[Tuple[AnyStr, AnyStr]]) -> Self: ...
```

We can't use PEP 673 Self yet, this is our hacked-up version that requires explicit `self` annotations.

**Comment by posita** [comment]
@JelleZijlstral, howTF did you see, much less have time to propose a change to this PR within 90 seconds of its creation?!

Shocked!

**Comment by hauntsaninja** [comment]
cat!

**Comment by JelleZijlstra** [comment]
I guess sometimes I am too bothered by the little blue box at the top right :)

File: stdlib/os/__init__.pyi [hunk 2]
```diff
@@ -245,6 +246,13 @@ class _Environ(MutableMapping[AnyStr, AnyStr], Generic[AnyStr]):
     def __setitem__(self, key: AnyStr, value: AnyStr) -> None: ...
     def __iter__(self) -> Iterator[AnyStr]: ...
     def __len__(self) -> int: ...
+    if sys.version_info >= (3, 9):
+        # We use @overload instead of a Union for reasons similar to those given for
+        # overloading MutableMapping.update in stdlib/typing.pyi
+        @overload
+        def __ior__(self: Self, __value: Mapping[AnyStr, AnyStr]) -> Self: ...
+        @overload
+        def __ior__(self: Self, __value: Iterable[Tuple[AnyStr, AnyStr]]) -> Self: ...
```
Comment (hauntsaninja) [inline stdlib/os/__init__.pyi:255]:
```suggestion
        def __ior__(self: Self, __value: Iterable[tuple[AnyStr, AnyStr]]) -> Self: ...
```

File: stdlib/os/__init__.pyi [hunk 3]
```diff
@@ -30,6 +30,7 @@ from typing import (
     NoReturn,
     Protocol,
     Sequence,
+    Tuple,
```
Comment (hauntsaninja) [inline stdlib/os/__init__.pyi:33]:
```suggestion
```

File: stdlib/os/__init__.pyi (hunk 2 above)
Comment (JelleZijlstra) [inline stdlib/os/__init__.pyi:255]: You should use lowercase `tuple` (I think that's why CI is failing)

File: stdlib/os/__init__.pyi [hunk 4]
```diff
@@ -245,6 +245,13 @@ class _Environ(MutableMapping[AnyStr, AnyStr], Generic[AnyStr]):
     def __setitem__(self, key: AnyStr, value: AnyStr) -> None: ...
     def __iter__(self) -> Iterator[AnyStr]: ...
     def __len__(self) -> int: ...
+    if sys.version_info >= (3, 9):
+        # We use @overload instead of a Union for reasons similar to those given for
+        # overloading MutableMapping.update in stdlib/typing.pyi
+        @overload
+        def __ior__(self: Self, __value: Mapping[AnyStr, AnyStr]) -> Self: ...
+        @overload
+        def __ior__(self: Self, __value: Iterable[tuple[AnyStr, AnyStr]]) -> Self: ...
```
Comment (JelleZijlstra) [inline stdlib/os/__init__.pyi:254]:
```suggestion
        def __ior__(self: Self, value: Mapping[AnyStr, AnyStr]) -> Self: ...
        @overload
        def __ior__(self: Self, value: Iterable[tuple[AnyStr, AnyStr]]) -> Self: ...
```
Apparently it's not pos-only at runtime

**Pull request merged**